ql/src/java/org/apache/hadoop/hive/ql/optimizer/BucketVersionPopulator.java

```diff
@@ -0,0 +1,248 @@
+    //    SemanticGraphWalker ogw = new PreOrderWalker(disp);
+    SemanticGraphWalker ogw = new DefaultGraphWalker(disp);
+
+    ArrayList<Node> topNodes = new ArrayList<Node>();
+    topNodes.addAll(pGraphContext.getTopOps().values());
+    ogw.startWalking(topNodes, null);
+    return pGraphContext;
+  }
+
+  class OpGroup {
+    Set<Operator<?>> members = Sets.newIdentityHashSet();
+    int version = -1;
+
+    public OpGroup() {
+      groups.add(this);
+    }
+
+    public void add(Operator o) {
+      members.add(o);
+      b.put(o, this);
+    }
+
+    public void setBucketVersion() {
+      for (Operator<?> operator : members) {
+        operator.getConf().setBucketingVersion(version);
+      }
+    }
+
+    class OperatorBucketingVersionInfo {
+
+      private Operator<?> op;
+      private int bucketingVersion;
+
+      public OperatorBucketingVersionInfo(Operator<?> op, int bucketingVersion) {
+        this.op = op;
+        this.bucketingVersion = bucketingVersion;
+      }
+
+      @Override
+      public String toString() {
+        return String.format("[op: %s, bucketingVersion=%d]", op, bucketingVersion);
+      }
+    }
+
+    List<OperatorBucketingVersionInfo> getBucketingVersions() {
+      List<OperatorBucketingVersionInfo> ret = new ArrayList<>();
+      for (Operator<?> operator : members) {
+        if (operator instanceof TableScanOperator) {
+          TableScanOperator tso = (TableScanOperator) operator;
+          int bucketingVersion = tso.getConf().getTableMetadata().getBucketingVersion();
+          int numBuckets = tso.getConf().getNumBuckets();
+          if (numBuckets > 1) {
+            ret.add(new OperatorBucketingVersionInfo(operator, bucketingVersion));
+          } else {
+            LOG.info("not considering bucketingVersion for: %s because it has %d<2 buckets ", tso, numBuckets);
+          }
+        }
+        if (operator instanceof FileSinkOperator) {
+          FileSinkOperator fso = (FileSinkOperator) operator;
+          int bucketingVersion = fso.getConf().getTableInfo().getBucketingVersion();
+          ret.add(new OperatorBucketingVersionInfo(operator, bucketingVersion));
+        }
+      }
+      return ret;
+    }
+
+    public void analyzeBucketVersion() {
+      List<OperatorBucketingVersionInfo> bucketingVersions = getBucketingVersions();
+      try {
+        for (OperatorBucketingVersionInfo info : bucketingVersions) {
+        setVersion(info.bucketingVersion);
+      }
+      } catch (Exception e) {
+        throw new RuntimeException("Error setting bucketingVersion for group: " + bucketingVersions, e);
+      }
+      if (version == -1) {
+        // use version 2 if possible
+        version = 2;
+      }
+    }
+
+    private void setVersion(int newVersion) {
+      if (version == newVersion || newVersion == -1) {
+        return;
+      }
+      if (version == -1) {
+        version = newVersion;
```

Comment:
Should we add in debug level a line here showing that we are setting the version for a certain op?